python/cudf_polars/cudf_polars/utils/config.py

```diff
@@ -5,209 +5,148 @@
+    """Configuration for the cudf-polars in-memory executor."""
 
-        Raises
-        ------
-        ValueError
-            If the configuration contains unsupported options.
-        """
-        if unsupported := (
-            config.keys()
-            - {"raise_on_fail", "parquet_options", "executor", "executor_options"}
-        ):
-            raise ValueError(
-                f"Engine configuration contains unsupported settings: {unsupported}"
-            )
-        assert {"chunked", "chunk_read_limit", "pass_read_limit"}.issuperset(
-            config.get("parquet_options", {})
-        )
+    name: Literal["in-memory"] = "in-memory"
+    scheduler: Literal["synchronous"] = "synchronous"
+    shuffle_method: Literal["tasks"] | None = None
+    broadcast_join_limit: int = 32
 
-        # Validate executor_options
-        executor = config.get("executor", "in-memory")
-        if executor == "streaming":
-            executor_options, unsupported = _valid_streaming_options(
-                config["executor_options"]
-            )
-            config["executor_options"] = executor_options
-        else:
-            unsupported = config.get("executor_options", {}).keys()
-        if unsupported:
-            raise ValueError(
-                f"Unsupported executor_options for {executor}: {unsupported}"
-            )
 
-        return config
-
-
-def _valid_option(
-    options: MutableMapping[str, Any],
-    key: str,
-    *,
-    default: Any,
-    choices: Sequence[Any] | None = None,
-    astype: type | None = None,
-) -> MutableMapping[str, Any]:
-    """Return a valid configuration option."""
-    value = options.get(key, default)
-    if astype is not None:
-        value = astype(value)
-    if choices is not None and value not in choices:
-        raise ValueError(f"Unsupported {key} option: {value}. Choices are: {choices}")
-    options[key] = value
-    return options
-
-
-def _valid_streaming_options(
-    options: MutableMapping[str, Any],
-) -> tuple[MutableMapping[str, Any], set[str]]:
-    """Return valid options for the streaming executor."""
-    default: Any
-    choices: Sequence[Any]
-    for key in (
-        _STREAMING_OPTIONS := (
-            "scheduler",  # Validate scheduler first
-            "fallback_mode",
-            "max_rows_per_partition",
-            "parquet_blocksize",
-            "cardinality_factor",
-            "groupby_n_ary",
-            "broadcast_join_limit",
-            "shuffle_method",
+@dataclasses.dataclass(frozen=True, eq=True)
+class ConfigOptions:
+    """Configuration for the polars GPUEngine."""
+
+    raise_on_fail: bool = False
+    # device?
+    # memory resource?
+    parquet_options: ParquetOptions = dataclasses.field(default_factory=ParquetOptions)
+    executor: StreamingExecutor | InMemoryExecutor = dataclasses.field(
+        default_factory=InMemoryExecutor
+    )
+
+    @classmethod
+    def from_polars_engine(cls, engine: pl.GPUEngine) -> Self:
```

Comment:
Note to self: see if we can handle https://github.com/rapidsai/cudf/pull/18531/files#diff-2f8f881e6eb5299e80fc0808df3d453617d6618fa7a51f1e8fd76cb52f2250b0R294 here.

We can check if `engine.memory_resource is None and executor == "streaming" and scheduler == "distributed"`.

That does mean we'd maybe have two `memory_resource` objects floating around: the user-provided `memory_resource` on `pl.GPUEngine(..., memory_resource=...)` and this other one, which would presumably be a new attribute of `ConfigOptions`.

Reply:
We need to think about how to handle memory resources in the distributed setting. We need a way for the user to specify the mr to be used on the workers, but we can't get them to pass an already initialised mr because you can't ship that around. We could ask them to provide a callable that would construct an mr though.

Reply:
A callable that returns a MemoryResource (by either constructing it or getting an existing one) makes sense...

We'll need to coordinate this with the `memory_resource` parameter provided to `polars.GPUEngine`, which I believe is only sensible for the `in-memory` executor. Maybe best as a followup?

Reply:
>We need to think about how to handle memory resources in the distributed setting.

Right now, we definitely don't want cudf-polars to be responsible for defining/creating/managing the memory resource used on the workers. In fact, we probably don't want the user to be allowed to use the `memory_resource` configuration at all. Cudf-polars is essentially a "client" in this case, and the user is responsible for deploying the cluster (with appropriate memory resources on each worker) separately. If/when we decide to make cluster deployment automagical, we will want those memory-resource options to be in some `DistributedOptions` configuration.